File: src/HeatFEM.m
classdef HeatFEM < matlab.mixin.Copyable
    %HEATFEM Summary of this class goes here
    %   Detailed explanation goes here
    
    properties
        gmsh
        spatialDimensions
        nodeCoordinates
        Dofs
        % All the information about the mesh is given above
        % The properties below are just helpers
        nbrNodes
        nbrDofs
        
        % Boundary conditions
        nbrBcs = 0;
        boundaryConditions = struct(...
            'physicalName', [], ...
            'type', [], ...
            'value', [], ...
            'timeSteps', [] ...
        );
        
        % Body conditions
        nbrBodyCond = 0;
        bodyConditions = struct(...
            'type', [], ...
            'physicalName', [] ...
        );
    
        appliedBoundaries = 0;
        appliedBodies = 0;
        
        % Default material parameters
        thickness = 1;
        rho = 1;
        cp = 1;
        D = [1 0; 0 1]; % conductivity matrix
        
        A % dR/dT_n matrix
        B % dR/dT_n-1 matrix
        
        K % stiffness matrix
        M % mass matrix
        Kc % Robin stiffness matrix
        fc % Robin load vector
        fl % Neumann load vector
        fv % Volume load vector
        
        % A cell array of size [1 x N] where N is the number of time steps
        % Entry nbr n represents the blocked degrees of freedom at time t_n
        blockedDofs
        
        transient = 0;
        temperatures
        loads
        
        tFinal = -1;
        timeSteps = 0;
        theta = 1; % Time integration variable
    end
    
    methods
        function obj = HeatFEM(gmshData, tFinal, timeSteps)
            %HEATFEM Construct an instance of this class
            %   Detailed explanation goes here
            obj.gmsh = gmshData;
            obj.nodeCoordinates = getGlobalCoordinates(gmshData);
            
            obj.nbrNodes = size(obj.nodeCoordinates, 1);
            
            % TODO: Generalize for vector fields
            obj.nbrDofs = obj.nbrNodes;
            obj.Dofs = (1:obj.nbrNodes)';
            
            % Initialize matrices
            obj.K = sparse(obj.nbrDofs, obj.nbrDofs);
            obj.Kc = sparse(obj.nbrDofs, obj.nbrDofs);
            obj.M = sparse(obj.nbrDofs, obj.nbrDofs);
            
            if exist('tFinal', 'var')
                obj.transient = 1;
                obj.tFinal = tFinal;
                if ~exist('timeSteps', 'var')
                    error(['The number of time steps must be specified if ' ...
                        'a transient model is chosem']);
                end
                obj.timeSteps = timeSteps;
            end
            
            % Initialize load vectors and temperature vector
            if obj.transient
                obj.fl = zeros(obj.nbrDofs, obj.timeSteps);
                obj.fv = zeros(obj.nbrDofs, obj.timeSteps);
                obj.fc = zeros(obj.nbrDofs, obj.timeSteps);
                obj.loads = zeros(obj.nbrDofs, obj.timeSteps);
                obj.temperatures = zeros(obj.nbrDofs, obj.timeSteps);
                % Create cell array as each entry of blocked dofs may vary in size
                obj.blockedDofs = cell(1, timeSteps);
            else
                obj.loads = zeros(obj.nbrDofs, 1);
                obj.temperatures = zeros(obj.nbrDofs, 1);
            end
        end
        
        function setMaterial(obj, material, thickness)
            obj.D = material.D;
            obj.rho = material.density;
            obj.cp = material.heatCapacity;

            obj.thickness = thickness;
        end
        
        function addBoundaryCondition(obj, boundaryCondition)
            obj.boundaryConditions(obj.nbrBcs + 1) = boundaryCondition;
            obj.nbrBcs = obj.nbrBcs + 1;
        end
        
        function addBodyCondition(obj, bodyCondition)
            obj.bodyConditions(obj.nbrBodyCond + 1) = bodyCondition;
            obj.nbrBodyCond = obj.nbrBodyCond + 1;
        end
        
        function assemble(obj)
            obj.applyBoundaryConditions();
            obj.applyBodyConditions();
            
            obj.loads = obj.fl + obj.fv + obj.fc;
            if obj.transient
                dt = obj.tFinal / (obj.timeSteps-1);
                obj.A = obj.M + dt*obj.theta*obj.K;
                obj.B = obj.M + dt*(obj.theta - 1)*obj.K;
                obj.loads = dt*obj.loads;
            end
        end
        
        function solve(obj)
            if obj.transient
                % f_bar for n = 1, 2, ..., N
                weightedLoads = obj.theta*obj.loads(:, 2:end) + ...
                    (1-obj.theta)*obj.loads(:, 1:end-1);

                obj.temperatures = obj.solveTransient(obj.temperatures, ...
                    weightedLoads, obj.blockedDofs);
            else
                obj.temperatures = obj.partitionAndSolve(obj.K, ...
                    obj.loads, obj.temperatures, obj.blockedDofs);
            end
        end
        
        function [Ex, Ey, elementTemp] = getElemTemp(obj, elementType, timeStep)
            Enod = getElements(obj.gmsh, elementType);
            xs = obj.nodeCoordinates(:, 1);
            ys = obj.nodeCoordinates(:, 2);
            Ex = xs(Enod(:, 2:end));
            Ey = ys(Enod(:, 2:end));
            
            temp = obj.temperatures(:, timeStep+1);
            elementTemp = temp(Enod(:, 2:end));
        end
        
        function conf = getConfiguration(obj)
            conf = struct(...
                'thickness', obj.thickness, ...
                'rho', obj.rho, ...
                'cp', obj.cp, ...
                'D', obj.D, ...
                'boundaryConditions', obj.boundaryConditions, ...
                'bodyConditions', obj.bodyConditions, ...
                'mesh', obj.gmsh, ...
                'tFinal', obj.tFinal', ...
                'timeSteps', obj.timeSteps, ...
                'theta', obj.theta ...
            );
        end
    end
    
    methods(Access = protected)
        function applyBoundaryConditions(obj)
            if ~obj.appliedBoundaries
                for bc = obj.boundaryConditions
                    elementBlocks = getPhysicalEntity(obj.gmsh, bc.physicalName);
                    switch bc.type
                        case 'Neumann'
                            magnitude = bc.value;
                            func = @(ex, ey, elementType) obj.elementLoad(...
                                ex, ey, elementType, magnitude, obj.thickness);
                            f = obj.integrate(elementBlocks, func, 1);
                            obj.fl(:, bc.timeSteps) = ...
                                obj.fl(:, bc.timeSteps) + ...
                                repmat(f, [1 length(bc.timeSteps)]);
                        case 'Robin'
                            magnitude = bc.alpha * bc.Tinf;
                            func = @(ex, ey, elementType) obj.elementLoad(...
                                ex, ey, elementType, magnitude, obj.thickness);
                            f = obj.integrate(elementBlocks, func, 1);
                            obj.fc(:, bc.timeSteps) = ...
                                obj.fc(:, bc.timeSteps) + ...
                                repmat(f, [1 length(bc.timeSteps)]);

                            magnitude = bc.alpha;
                            func = @(ex, ey, elementType) obj.elementMass(...
                                ex, ey, elementType, magnitude, obj.thickness);
                            % TODO: handle different Kc at different time steps
                            obj.Kc = obj.Kc + obj.integrate(elementBlocks, func, 2);

                        case 'Dirichlet'
                            % Extract all the nodes of the element blocks
                            nodes = [];
                            for elementBlock = elementBlocks
                                nodes = horzcat(nodes, [elementBlock.elements.nodeTags]);
                            end
                            bd = obj.Dofs(unique(nodes));
                            addBd = @(c) {[c, bd]};
                            obj.blockedDofs(bc.timeSteps) = cellfun(addBd, ...
                                obj.blockedDofs(bc.timeSteps));
                            obj.temperatures(bd, bc.timeSteps) = bc.value;
                    end
                end
                obj.appliedBoundaries = 1;
            end
        end
        
        function applyBodyConditions(obj)
            if ~obj.appliedBodies
                for bc = obj.bodyConditions
                    elementBlocks = getPhysicalEntity(obj.gmsh, bc.physicalName);
                    switch bc.type
                        case 'main'
                            func = @(ex, ey, elementType) obj.elementStiffness(...
                                ex, ey, elementType, obj.D, obj.thickness);
                            obj.K = obj.K + obj.integrate(elementBlocks, func, 2);

                            magnitude = obj.rho * obj.cp;
                            func = @(ex, ey, elementType) obj.elementMass(...
                                ex, ey, elementType, magnitude, obj.thickness);
                            obj.M = obj.M + obj.integrate(elementBlocks, func, 2);
                        case 'load'
                            magnitude = bc.value;
                            func = @(ex, ey, elementType) obj.elementLoad(...
                                ex, ey, elementType, magnitude, obj.thickness);
                            f = obj.integrate(elementBlocks, func, 1);
                            obj.fv(:, bc.timeSteps) = ...
                                obj.fv(:, bc.timeSteps) + ...
                                repmat(f, [1 length(bc.timeSteps)]);
                        case 'initial'
                            % Extract all the nodes of the element blocks
                            nodes = [];
                            for elementBlock = elementBlocks
                                nodes = horzcat(nodes, [elementBlock.elements.nodeTags]);
                            end
                            bd = obj.Dofs(unique(nodes));
                            obj.temperatures(bd, 1) = bc.value;
                    end
                end
                obj.appliedBodies = 1;
            end
        end
        
        function [matrix] = integrate(obj, elementBlocks, func, dim)
            if dim == 1
                matrix = zeros(obj.nbrDofs, 1);
            elseif dim == 2
                matrix = zeros(obj.nbrDofs, obj.nbrDofs);
            else
                error('The input "dim" must be either 1 or 2');
            end
            for block = elementBlocks
                for element = block.elements
                    ex = obj.nodeCoordinates(element.nodeTags, 1);
                    ey = obj.nodeCoordinates(element.nodeTags, 2);
                    edof = obj.Dofs(element.nodeTags);
                    elemMatrix = func(ex, ey, block.elementType);
                    if dim == 1
                        matrix(edof) = matrix(edof) + elemMatrix;
                    elseif dim == 2
                        matrix(edof, edof) = matrix(edof, edof) + elemMatrix;
                    end
                end
            end
            matrix = sparse(matrix);
        end
        
        function X = solveTransient(obj, X, F, bds)
            % For each time step, solve the linear transient system AX=Y
            bd = bds{1};
            if isempty(bd)
                freeDofs = 1:obj.nbrDofs;
            else
                freeDofs = ~any(1:obj.nbrDofs == bd);
            end
            A = sparse(obj.A);
            % Decompose the A-matrix to speed up the linear system solver,
            % which prevents us from using the partitionAndSolve function
            % (Works only for linear systems, and for constant Dirichlet
            % boundary conditions)
            dA = decomposition(A(freeDofs, freeDofs));
            for timeStep = 1:(obj.timeSteps - 1)
                Yn = obj.B * X(:, timeStep) + F(:, timeStep);
                Xn = X(:, timeStep+1);
                bd = bds{timeStep};
                
                if isempty(bd)
                    freeDofs = 1:obj.nbrDofs;
                else
                    freeDofs = ~any(1:obj.nbrDofs == bd);
                end
                partYn = Yn(freeDofs) - ...
                        A(freeDofs, ~freeDofs)* ...
                        Xn(~freeDofs);
                
                partXn = dA \ partYn;
                Xn(freeDofs) = partXn;
                
                X(:, timeStep + 1) = Xn;
            end
        end
    end
    
    methods(Static)
        function [Ke] = elementStiffness(ex, ey, elementType, D, thickness)
            switch elementType
                case Elements.TRI_3
                    Ke = hTRI_3K(ex, ey, D, thickness);
                case Elements.QUA_4 % QUADS
                    Ke = flw2i4e(ex', ey', [thickness 2], D);
                    %Ke = hQUA_4K(ex, ey, obj.D, obj.thickness);
                otherwise
                    error("The element stiffness matrix is not yet implemented for the current element type");
            end
        end
        
        function [Me] = elementMass(ex, ey, elementType, rho, thickness)
            switch elementType
                case Elements.TRI_3
                    Me = hTRI_3M(ex, ey, rho, thickness);
                case Elements.QUA_4 % QUADS
                    Me = hQUA_4M(ex, ey, rho, thickness);
                otherwise
                    error("The element mass matrix is not yet implemented for the current element type");
            end
        end
        
        function [fe] = elementLoad(ex, ey, elementType, magnitude, thickness)
            switch elementType
                case Elements.PNT
                    fe = magnitude*thickness;
                case Elements.LIN_2
                    fe = hLIN_2f(ex, ey, magnitude, thickness);
                otherwise
                    error("The element load vector is not yet implemented for the current element type");
            end
        end
        
        function [X] = partitionAndSolve(A, Y, X, blockedDofs)
            % PARTITIONANDSOLVE Partition the system AX = Y and solve it
            freeDofs = ~any(1:length(X) == blockedDofs);
            partY = Y(freeDofs) - ...
                    A(freeDofs, ~freeDofs)* ...
                    X(~freeDofs);
            partA = A(freeDofs, freeDofs);
            partX = partA \ partY;
            X(freeDofs) = partX;
        end
    end
end

File: julia/HeatFEM.jl
include("GlobalAssembly.jl")
include("GmshParser.jl")
include("ElementRoutines.jl")

using SparseArrays
using LinearAlgebra

function node_to_dof(node_nbr)
    node_nbr
end

# Dispatch on element type

element_stiffness(::Type{Val{QUA_4}}, ex, ey, D, thickness) =
    hQUA_4K(ex, ey, D, thickness)

element_mass(::Type{Val{QUA_4}}, ex, ey, rho, thickness) =
    hQUA_4M(ex, ey, rho, thickness)

element_load(::Type{Val{LIN_2}}, ex, ey, load, thickness) =
    hLIN_2f(ex, ey, load, thickness)

element_load(::Type{Val{PNT}}, ex, ey, load, thickness) =
    load*thickness

struct BoundaryCondition
    type            :: String
    physical_name   :: String
    element_type    :: ElementType
    connectivity    :: Matrix{Int}
    value           :: Float64
    time_steps      :: Vector{Int}
end

function BoundaryCondition(g::Gmsh, type::AbstractString, physical_name::AbstractString,
    value::Float64, time_steps::AbstractArray{Int})
    element_type = get_element_type(g, physical_name)
    connectivity = get_elements(g, physical_name)
    return BoundaryCondition(type, physical_name, element_type, connectivity, value, time_steps)
end

struct BodyCondition
    type            :: String
    physical_name   :: String
    element_type    :: ElementType
    connectivity    :: Matrix{Int}
    value           :: Float64
end

BodyCondition(g::Gmsh, type::AbstractString, physical_name::AbstractString) =
    BodyCondition(g, type, physical_name, 0.0)

function BodyCondition(g::Gmsh, type::AbstractString, physical_name::AbstractString,
    value::Float64)
    element_type = get_element_type(g, physical_name)
    connectivity = get_elements(g, physical_name)
    return BodyCondition(type, physical_name, element_type, connectivity, value)
end


struct Material
    thickness       :: Float64
    density         :: Float64
    heat_capacity   :: Float64
    thermal_cond    :: Matrix{Float64}
end

mutable struct HeatFEM
    node_coordinates:: Matrix{Float64}
    t_final         :: Float64
    time_steps      :: Int
    theta           :: Float64
    nbr_nodes       :: Int
    nbr_dofs        :: Int
    temperatures    :: Matrix{Float64}
    K               :: SparseMatrixCSC{Float64, Int}
    C               :: SparseMatrixCSC{Float64, Int}
    F               :: Matrix{Float64}
    blocked_dofs    :: Vector{Vector{Int}}
    body_conditions :: Vector{BodyCondition}
    boundary_conditions :: Vector{BoundaryCondition}
    material        :: Material
end

function HeatFEM(node_coordinates::Matrix{Float64}, t_final::Int, time_steps::Int)
    return HeatFEM(node_coordinates, convert(Float64, t_final), time_steps)
end

function HeatFEM(node_coordinates::Matrix{Float64}, t_final::Float64, time_steps::Int)
    nbr_nodes = size(node_coordinates, 1)

    dofs = node_to_dof.(1:nbr_nodes)
    nbr_dofs = length(dofs)

    temperatures = zeros(nbr_dofs, time_steps)

    K = spzeros(nbr_dofs, nbr_dofs)
    C = spzeros(nbr_dofs, nbr_dofs)
    F = spzeros(nbr_dofs, time_steps)

    blocked_dofs = fill([], time_steps)
    body_conditions = Vector{BodyCondition}(undef, 0)
    boundary_conditions = Vector{BoundaryCondition}(undef, 0)

    theta = 1
    material = Material(1, 1, 1, Matrix{Float64}(I, 2, 2))

    return HeatFEM(node_coordinates, t_final, time_steps, theta, nbr_nodes, nbr_dofs,
        temperatures, K, C, F, blocked_dofs, body_conditions,
        boundary_conditions, material)
end

function get_matrices(obj::HeatFEM)
    dt = obj.t_final / (obj.time_steps - 1)
    A = obj.C + dt*obj.theta*obj.K
    B = obj.C + dt*(obj.theta - 1)*obj.K
    return A, B
end

function solve_transient!(X, A, B, F, bds)
    nbr_dofs, time_steps = size(X)
    bd = bds[1]
    free_dofs = setdiff(1:nbr_dofs, bd)
    dA = cholesky(A[free_dofs, free_dofs])
    for time_step = 1:time_steps-1
        Y = B * X[:, time_step] + F[:, time_step]
        partY = Y[free_dofs] - A[free_dofs, bd]*X[bd, time_step+1]
        X[free_dofs, time_step + 1] = dA \ partY
    end
end

function apply_body_conditions!(obj::HeatFEM)
    D = obj.material.thermal_cond
    s = obj.material.density * obj.material.heat_capacity
    thickness = obj.material.thickness

    K = GlobalAssembly(obj.nbr_dofs, obj.nbr_dofs)
    C = GlobalAssembly(obj.nbr_dofs, obj.nbr_dofs)
    F_v = spzeros(obj.nbr_dofs, obj.time_steps)
    initial_temp = zeros(obj.nbr_dofs, 1)

    for bc = obj.body_conditions
        if bc.type == "main"
            @time begin
            for e = 1:size(bc.connectivity, 1)
                enod = bc.connectivity[e, 2:end]
                edof = node_to_dof.(enod)
                ex = obj.node_coordinates[enod, 1]
                ey = obj.node_coordinates[enod, 2]
                Ke = element_stiffness(Val{bc.element_type}, ex, ey, D, thickness)
                Ce = element_mass(Val{bc.element_type}, ex, ey, s, thickness)

                insert!(K, edof, edof, Ke)
                insert!(C, edof, edof, Ce)
            end
            end

        elseif bc.type == "load"
            load = bc.value
            f = GlobalAssembly(obj.nbr_dofs)
            for e = 1:size(bc.connectivity, 1)
                enod = bc.connectivity[e, 2:end]
                edof = node_to_dof.(enod)
                ex = obj.node_coordinates[enod, 1]
                ey = obj.node_coordinates[enod, 2]
                fe = element_load(Val{bc.element_type}, ex, ey, load, thickness)

                insert!(f, edof, fe)
            end
            F_v[:, bc.time_steps] .+= f

        elseif bc.type == "initial"
            for e = 1:size(bc.connectivity, 1)
                enod = bc.connectivity[e, 2:end]
                edof = node_to_dof.(enod)
                initial_temp[dofs] = bc.value
            end
        end
    end

    obj.C = C
    obj.K = K
    obj.F += F_v
    obj.temperatures[:, 1] = initial_temp
end

function apply_boundary_conditions!(obj::HeatFEM)
    D = obj.material.thermal_cond
    s = obj.material.density * obj.material.heat_capacity
    thickness = obj.material.thickness

    F = spzeros(obj.nbr_dofs, obj.time_steps)
    for bc = obj.boundary_conditions
        if bc.type == "Neumann"
            load = bc.value
            f = GlobalAssembly(obj.nbr_dofs)
            for e = 1:size(bc.connectivity, 1)
                enod = bc.connectivity[e, 2:end]
                edof = node_to_dof.(enod)
                ex = obj.node_coordinates[enod, 1]
                ey = obj.node_coordinates[enod, 2]
                fe = element_load(Val{bc.element_type}, ex, ey, load, thickness)

                insert!(f, edof, fe)
            end
            f = convert(SparseMatrixCSC, f)
            F[:, bc.time_steps] .+= f

        elseif bc.type == "Robin"
            @error "The robin boundary conditions are not yet implemented"

        elseif bc.type == "Dirichlet"
            nodes = unique(bc.connectivity[:, 2:end])
            for time_step in bc.time_steps
                append!(obj.blocked_dofs[time_step], nodes)
            end
            obj.temperatures[nodes, bc.time_steps] .= bc.value
        end
    end

    obj.F += F
end

function solve!(obj::HeatFEM)
    dt = obj.t_final / (obj.time_steps - 1)
    weightedLoads = dt*(obj.theta * obj.F[:, 2:end] + (1 - obj.theta)*obj.F[:, 1:end-1])
    A, B = get_matrices(obj)
    solve_transient!(obj.temperatures, A, B, weightedLoads, obj.blocked_dofs)
end
